components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt

```diff
@@ -16,41 +17,22 @@ class ToolbarPresenter(
     private val toolbar: Toolbar,
     private val sessionManager: SessionManager,
     private val sessionId: String? = null
-) : SessionManager.Observer, Session.Observer {
-    private var activeSession: Session? = null
+) : SessionObserver(sessionManager) {
 
     /**
      * Start presenter: Display data in toolbar.
      */
     fun start() {
-        activeSession = sessionId?.let {
-            sessionManager.findSessionById(sessionId)
-        } ?: run {
-            sessionManager.register(this)
-            sessionManager.selectedSession
-        }
-
-        activeSession?.register(this)
+        val session = sessionId?.let { sessionManager.findSessionById(sessionId) }
+        session?.let { observe(it) } ?: run { observeSelected() }
```

Comment:
Is the `run {}` block really needed here?

Reply:
No, it's not needed. I just found it easier to read `let {} ?: run {}`, but I'll remove it if it's confusing.